### ISSUE TITLE
Document handling questions

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#559]
@damirius @FWAJL 
I have created a new github Milestone "Document Handling".  Note that github Milestones have nothing to do with eLance Milestones - this was a bit confusing to me at first!

The github Milestone "Document Handling" groups all Issues related to document handling.  So you could look at the communication history for each of the related issues to help you get oriented.

This is what has been done:
* The table "document" has been added to the database.
* The Dao class "document.php" has been created.
* The appsettings key "RootDocumentpload" has been defined.
* The view module "upload_file" has been started.

We will need to handle, primarily, two types of documents: photos and forms.  Photos can have several extensions (jpg, png, etc.).  Forms shall be pdf files.  For mapping, we only need photos, but we are building the controller to handle the pdf forms also.

Jeremie may have specific ideas for how to mange document uploads and views, so you may need to communicate with him.  I don't have a preference as long as it is a cost effective and efficient way to fulfill my needs.

**Comment by damirius** [comment]
@bsaiken 

Thanks, I'll take a closer look at those issues first. Since @FWAJL said that he will review my mapping Milestone B code at the start of the next week I'll probably work on this over the weekend so we can start using it before I hit Milestone D for mapping.

Thanks again

**Comment by bsaiken** [comment]
@damirius 
If you you are going to spend any amount of time on it, I would want to make formal arrangements, so I could pay you.  If we did it as a minor change of scope within the mapping project, that is easy as long as it is in $50 increments.  Just let me know what you want to do.  I will be available and working on the site all weekend.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

The code I have done so far on document upload is showcased when editing a technician. I have not completed the whole thing but the upload part is almost complete.

Adding documents should be straight forward and almost nothing needs to be added. To test from end to end.

Now, we need a FileLoader (FieldWorkAssistantMVC\Library\Core\Utility\FileLoader.php) to read the database for the file(s) info that will allow to find them in the uploads folder.

The naming of the files is [ID_OF_DOCUMENT_CATEGORY]_[UNIQUE_ID].[EXTENSION]
For ex, a picture for technician (ID=1) will be ***1***_***0d37ac83-94c8-4b7e-ae47-42f197380b8e***.***jpeg***.

When reading the database for all the documents for technician (ID=1), we need to select all the rows where document_category is technician_id and document_value starts with "1_"
See the attached image of how it looks in the database. You can try this locally.
![image 2](https://cloud.githubusercontent.com/assets/7524388/6324717/471f64c8-bb3e-11e4-8768-96ae1620fe9b.png)


I will try to answer more questions later. Let me know if you understand the upload first.

Thank you.

**Comment by FWAJL** [comment]
@damirius 

FYI, I have chosen to build the document value to be unique but if you have another idea or an experience to share, please do.

Thank you.

**Comment by damirius** [comment]
@FWAJL 

All clear. I don't see a problem with assigning unique value to a document name.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

Are we all clear? Is there more clarifications to give?

Thank you.

**Comment by damirius** [comment]
@FWAJL 

Hey, I have a question regarding object serialization. I know this isn't proper place for that but I ran into some problems while working on document handling so I wanted to ask you here. If you think it's OK we can open another Issue where we will discuss this problem further.

As you know, I've implemented `IDocument` interface for `Entities` which contain files. So I had to add two new members, `webPath` and `absolutePath` to each `IDocument` class. I had to set them to private since our `BaseManager` is using public members to create queries. So when we do `json_encode` on those objects it will not encode everything that we need.
I could add it to the IDocument interface and implement it in classes that use the mentioned interface, but I think this should be implemented somewhere in `Entity` abstract class since we could use it on other type of entity objects as well.
We just need some small method which we can use to specify which members to return.
For example `GetSerializable()` which will contain all members that we need to return as json later. Or we can even pass it some variable to see if we need to return, for example, just `id` and `name` or something more than that.
After that we can call that method for each object and send it like that. 
We can also use http://php.net/manual/en/jsonserializable.jsonserialize.php.

Thanks!
